.travis.yml

```diff
@@ -1,133 +1,192 @@
+    "${PYENV}" global ${PYENV_VERSION}
+    pip install -U tox "setuptools>=36"
+
 x-linux-shard: &x-linux-shard
+  <<: *x-pyenv-shard
   os: linux
   sudo: false
   dist: trusty
-  language: python
-  cache:
-    directories:
-      - .pyenv_test
 
-x-osx-shard: &x-osx-shard
-  os: osx
-  language: generic
+# Python 3.7 requires at least OpenSSL 1.0.2:
+#   https://docs.python.org/3/whatsnew/3.7.html#platform-support-removals.
+# Travis' `trusty` image does not get us there and, at the time of writing, the beta `xenial`
+# distribution is not generally available. As such, we install and activate a custom OpenSSL
+# here - the latest 1.0.2 version at the time of writing.
+x-linux-37-shard: &x-linux-37-shard
+  <<: *x-linux-shard
+  addons:
+    apt:
+      packages:
+        # Even though we install our own, it does not ship with a root CA store, so we ensure one
+        # here at /usr/lib/ssl/certs and activate it by exporting SSL_CERT_DIR which OpenSSL knows
+        # how to read.
+        - openssl
+  env:
+    - *env
+    - OPENSSL_VERSION=1.0.2p
+    - OPENSSL_DIR="${HOME}/.pyenv_pex_openssl"
+    - LD_LIBRARY_PATH="${OPENSSL_DIR}/lib"
+    - SSL_CERT_DIR=/usr/lib/ssl/certs
+    - PYTHON_CONFIGURE_OPTS="--with-openssl=${OPENSSL_DIR}"
+    - PYENV_VERSION=3.7.0
+    - TOX_TESTENV_PASSENV=SSL_CERT_DIR
   cache:
+    <<: *cache
     directories:
       - .pyenv_test
-      - ${HOME}/.pyenvpex
+      - "${PYENV_ROOT}"
+      - "${OPENSSL_DIR}"
   before_install: |
-    export PYENV_ROOT="${HOME}/.pyenvpex"
-    if [ ! -d "${PYENV_ROOT}" ]; then
-      git clone https://github.com/pyenv/pyenv "${PYENV_ROOT}";
+    if [ ! -f "${OPENSSL_DIR}/installed-${OPENSSL_VERSION}" ]; then
+      rm -rf "${OPENSSL_DIR}"
+      curl https://www.openssl.org/source/openssl-${OPENSSL_VERSION}.tar.gz | tar -xz
+      (
+        cd openssl-${OPENSSL_VERSION}
+        ./config shared --prefix="${OPENSSL_DIR}"
+        make
+        make install
+      )
+      touch "${OPENSSL_DIR}/installed-${OPENSSL_VERSION}"
     fi
-    export PATH="${PYENV_ROOT}/bin:${PATH}"
-    hash -r
-    pyenv install --keep --skip-existing ${PYENV_VERSION}
-    pyenv global ${PYENV_VERSION}
 
-install:
-  - pip install -U tox "setuptools>=36"
+x-osx-shard: &x-osx-shard
+  <<: *x-pyenv-shard
+  os: osx
+  osx_image: xcode9.4
 
+x-py27: &x-py27
+  - *env
+  - PYENV_VERSION=2.7.15
+
+x-py37: &x-py37
+  - *env
+  - PYENV_VERSION=3.7.0
+
+x-pypy: &x-pypy
+  - *env
+  - PYENV_VERSION=pypy2.7-6.0.0
+
+# NB: Travis partitions caches using a combination of os, language amd env vars. As such, we do not
+# use TOXENV and instead pass the toxenv via -e on the command line. This helps ensure we share
+# caches as much as possible (eg: all linux python 2.7.15 shards share a cache).
 matrix:
   include:
     - <<: *x-linux-shard
-      python: "2.7"
       name: TOXENV=style
-      script: ./scripts/ci.sh style
+      env: *x-py27
+      script: tox -ve style
```

Comment:
Are there long flag names for this? I'm not sure what `-v` and `-e` are supposed to mean.

Reply:
Yeah - I'll roll this up into an env var. I was avoiding doin that just because changing an env var kills the caches and the shards go from 4 minutes to ~20. Better to do that now though before this hits master to seal a good re-usable master cache set in for future PRs.

Reply:
No long opts for these, but fixed with a documented factor-up.